xpar.1
.TH XPAR 1
.SH NAME
xpar \- an error/erasure code system guarding data integrity
.SH SYNOPSIS
.ll +8
.B xpar
.RB [ " \-Je " / " \-Jd " ]
.RB [ " \-hfvqVc " ]
.RB [ " \-i/--interlacing\ # " ]
.RB [ " \--no-mmap\ " ]
[
.I "names \&..."
]
.ll -8
.br
.B xpar
.RB [ " \-Se " ]
.RB [ " \-hfvqV " ]
.RB [ " \--out-prefix\ # ", " \--dshards\ # ", " \--pshards\ # " ]
.RB [ " \--no-mmap\ " ]
[
.I "names \&..."
]
.br
.B xpar
.RB [ " \-Sd " ]
.RB [ " \-hfvqV " ]
[
.I "names \&..."
]
.SH DESCRIPTION
.PP
The
.B xpar
command
is the command-line interface to an error/erasure code system which
fortifies data and corrects errors in the data or reconstructs missing
shards.
.PP
.B xpar
in joint encoding mode (
.B \-Je
) generates a slightly inflated (by about 12%) parity-guarded
file from a given data file with the extension
.I .xpar
. Such a file can be recovered as long as no more than about 6.2% of the
data is corrupted. xpar internally uses a (255,223)-RS code over an 8-bit
Galois field. At most two filenames intertwined with flags are expected
as arguments.
.B xpar
will by default not overwrite existing files. If this behaviour is intended,
use the
.B \-f
option.
If no file names are specified,
.B xpar
will process data from the standard input and write it to the standard output,
refusing to output binary data to a terminal. If two files are specified,
the first one is used in place of standard input, and the second one is used
in place of standard output. If the -c flag is present,
.B xpar
will read from the specified file and output data to standard output instead.
Otherwise, if decoding, 
.B xpar
will try to guess the decompressed filename by removing the
.I .xpar
extension. If not present, an error will be reported. If encoding
the output filename will be generated by appending the
.I .xpar
extension to the input filename.
.PP
In joint mode, the flag
.B \-i/\-\-interlacing
can be used to specify the interlacing factor. This does not affect the nominal
error correction capability (in the sense of a noisy memory-less channel model),
but can be used to increase the robustness against burst errors. The default
interlacing factor is 1 and the maximum correctible burst error length is
16 (as given by the Singleton bound). The interlacing factor of 2 transposes
65025-byte blocks of data and hence can correct burst errors of length 4080. The
interlacing factor of 3 transposes approx. 16MB blocks of data and can correct
burst errors of length approx. 1MB. In theory, with the interlacing factor of
one, up to 32 burst errors could be corrected - this problem is however NP-hard
and reduces to polynomial reconstruction (the basis of e.g. Shamir's secret
sharing scheme). As an added benefit, higher interlacing factors tend to result
in faster processing (up to 50%) as the workload is more parallelisable.
.PP
.B xpar
in sharded encoding mode (
.B \-Se
) generates a set of parity-guarded files from a given data file. The amount of
files generated is given by the sum of the mandatory parameters
.B \--dshards
and
.B \--pshards
- which specify the number of data and parity shards, respectively. The data
shards contain the file contents verbatim (
.B xpar\ 
uses a systematic code). The data is reconstructible as long as there are more
than
.B \--dshards
shards available.
.SH OPTIONS
.TP
.B \-S --sharded
Enable sharded processing mode.
.TP
.B \-J --joint
Enable joint processing mode.
.TP
.B \-e --encode
Enable encoding mode.
.TP
.B \-d --decode
Enable decoding mode.
.TP
.B \-h --help
Print a help message and exit.
.TP
.B \-f --force
Force operation. Ignore errors and try to recover from them, if possible.
Overwrite existing files.
.TP
.B \-v --verbose
Be verbose. Print additional information.
.TP
.B \-q --quiet
Be quiet. Suppress most output, except fatal errors.
.TP
.B \-V --version
Print version information of the package and exit.
.TP
.B \-c --stdout
Write resultant data to standard output. Joint mode only.
.TP
.B \-i --interlacing
Specify the interlacing factor. Joint mode only.
.TP
.B \--no-mmap
Disable memory mapping. Generally results in worse performance, as the fallback
method is to read the file in chunks using the standard I/O library.
.TP
.B \-j --jobs
Specify the amount of CPU cores to use. Not setting this value or setting it to
zero will result in the program automatically deciding the amount of cores to
use. Setting it to one will disable parallel processing.
.TP
.B \--out-prefix
Specify the prefix for the output files (shards). Sharded mode only.
.TP
.B \--dshards
Specify the number of data shards. Sharded mode only. Mandatory.
.TP
.B \--pshards
Specify the number of parity shards. Sharded mode only. Mandatory.
.SH REPORTING BUGS
Report bugs to: k@iczelia.net or open a ticket
at https://github.com/kspalaiologos/xpar.
.SH COPYRIGHT NOTICE
Copyright \(co 2022-2024 Kamila Szewczyk
.PP
Thanks to: Phil Karn (KA9Q), for the original implementation of the Reed-Solomon
code, licensed under the GPLv2-or-later license. Marshall Lochbaum for noticing
an optimisation to the syndrome computaton code. Rich Felker for the
implementation of
.B getopt_long
in musl libc that is included in this package, licensed under the terms of the
Expat/MIT license.
.PP
Permission is granted to make and distribute verbatim copies of
this manual provided the copyright notice and this permission notice
are preserved on all copies. Permission is granted to process this file through troff and print the
results, provided the printed document carries copying permission
notice identical to this one except for the removal of this paragraph
(this paragraph not being relevant to the printed manual). Permission is granted to copy and distribute modified versions of this
manual under the conditions for verbatim copying, provided that the entire
resulting derived work is distributed under the terms of a permission
notice identical to this one. Permission is granted to copy and distribute translations of this manual
into another language, under the above conditions for modified versions,
except that this permission notice may be stated in a translation approved
by the Foundation.


NEWS
This file contains release notes for major and minor releases of xpar.
For a complete list of source-level changes, consult the ChangeLog file.

===============================================================================
v0.5 (17-10-2024)
- OpenMP support for sharded mode (which unfortunately seems bottlenecked by
  I/O). 
- Switch to yarg for command-line parsing, remove dependency on Rich Felker's
  `getopt_long`.
- Hopefully the last v0.x release. Hopefully, it will receive some feedback
  which will help to introduce future improvements and release v1.0. The file
  format will not change from now on, unless there is a bug or another major
  misfeature that needs to be fixed.

===============================================================================
v0.4 (16-10-2024)
- x86_64 static Linux binaries are no longer provided.
- OpenMP support has been added to improve encoding and decoding performance
  in joint mode with high interlacing factors on multi-core machines.
- 3-way saturating CRC32C implementation has been added to improve performance
  on x86_64 machines that support SSE4.2.
- Slightly improve the performance of the sharded mode.
- Fix undefined behaviour in sharded mode regarding int shifts.

===============================================================================
v0.3 (16-10-2024)
- Improve joint encoding performance on x86_64 machines.
- Support aarch64 Linux.
- Improve cross-platform compatibility of the sharded mode.

===============================================================================
v0.2 (15-10-2024)
- Provides a manual page for the xpar command.
- Provides platform-specific code for aarch64, which can be enabled via
  the --enable-aarch64 configure option.

===============================================================================
v0.1 (14-10-2024)
- Initial release.
- Supports joint mode and sharded mode for error and erasure correction.
- Provides platform-specific code for x86_64, which can be enabled via
  the --enable-x86_64 configure option.
- Tested on x86_64 Linux (Ubuntu), x86_64 and aarch64 MacOS and x86_64 and
  i686 Windows.


xpar.c
/*
   Copyright (C) 2022-2024 Kamila Szewczyk

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 3 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program. If not, see <http://www.gnu.org/licenses/>.
*/

#include "config.h"
#include "common.h"
#include "jmode.h"
#include "smode.h"
#include "platform.h"
#include "crc32c.h"
#include "yarg.h"

#include <limits.h>
#include <assert.h>
#include <stdint.h>
#include <errno.h>
#include <fcntl.h>
#include <sys/time.h>
#include <sys/stat.h>
#include <sys/types.h>

#if defined(XPAR_OPENMP)
#include <omp.h>
#endif

// ============================================================================
//  Command-line stub.
// ============================================================================
static void version() {
  printf(
    "xpar %d.%d. Copyright (C) by Kamila Szewczyk, 2022-2024.\n"
    "Licensed under the terms of GNU GPL version 3, available online at:\n"
    " <https://www.gnu.org/licenses/gpl-3.0.en.html>\n"
    "This is free software: you are free to change and redistribute it.\n"
    "There is NO WARRANTY, to the extent permitted by law.\n",
    XPAR_MAJOR, XPAR_MINOR
  );
}
static void help() {
  printf(
    "xpar - an error/erasure code system guarding data integrity.\n"
    "Usage (joint mode):\n"
    "  xpar -Je/-Jd [...] <in>              (adds/removes .xpa in output)\n"
    "  xpar -Je/-Jd [...] <in> <out>        (produces <out>)\n"
    "Usage (sharded mode):\n"
    "  xpar -Se [...] <in>                  (produces <in>.xpa.XXX)\n"
    "  xpar -Se --out-prefix=# [...] <in>   (produces #.xpa.XXX)\n"
    "  xpar -Sd [...] <out> <in>.001 ...    (produces <out>)\n"
    "\n"
    "Mode selection:\n"
    "  -J,   --joint        use the joint mode (default)\n"
    "  -S,   --sharded      use the sharded mode\n"
    "  -e,   --encode       add parity bits to a specified file\n"
    "  -d,   --decode       recover the original data\n"
    "Options:\n"
    "  -h,   --help         display an usage overview\n"
    "  -f,   --force        force operation: ignore errors, overwrite files\n"
    "  -v,   --verbose      verbose mode (display more information)\n"
    "  -q,   --quiet        quiet mode (display less information)\n"
    "  -V,   --version      display version information\n"
#if defined(XPAR_ALLOW_MAPPING)
    "        --no-mmap      unconditionally disable memory mapping\n"
#endif
#if defined(XPAR_OPENMP)
    "  -j #, --jobs=#       set the number of threads to use\n"
#endif
    "Joint mode only:\n"
    "  -c,   --stdout       force writing to standard output\n"
    "  -i #, --interlace=#  change the interlacing setting (1,2,3)\n"
    "Sharded mode encoding options:\n"
    "        --dshards=#    set the number of data shards (< 128)\n"
    "        --pshards=#    set the number of parity shards (< 64)\n"
    "        --out-prefix=# set the output file prefix\n"
    "\n"
    "In joint mode:\n"
    "Interlacing transposes the input data into larger blocks allowing\n"
    "the codes to accomplish higher burst error correction capabilities.\n"
    "The default interlacing factor is 1, which means no interlacing.\n"
    "The interlacing factor of two allows correction of 4080 contiguous\n"
    "errors in a 65025 byte block.\n"
    "\n"
    "Report bugs to: https://github.com/kspalaiologos/xpar\n"
    "Or contact the author: Kamila Szewczyk <k@iczelia.net>\n"
  );
}
enum mode_t { MODE_NONE, MODE_ENCODING, MODE_DECODING };
int main(int argc, char * argv[]) {
  jmode_gf256_gentab(0x87);  smode_gf256_gentab(0x87);
  platform_init();
  enum { FLAG_NO_MMAP = CHAR_MAX + 1, FLAG_DSHARDS, FLAG_PSHARDS,
         FLAG_OUT_PREFIX };
  yarg_options opt[] = {
    { 'V', no_argument, "version" },
    { 'v', no_argument, "verbose" },
    { 'J', no_argument, "joint" },
    { 'S', no_argument, "sharded" },
#if defined(XPAR_OPENMP)
    { 'j', required_argument, "jobs" },
#endif
    { 'c', no_argument, "stdout" },
    { 'q', no_argument, "quiet" },
    { 'h', no_argument, "help" },
    { 'e', no_argument, "encode" },
    { 'd', no_argument, "decode" },
    { 'f', no_argument, "force" },
    { FLAG_DSHARDS, required_argument, "dshards" },
    { FLAG_PSHARDS, required_argument, "pshards" },
    { FLAG_OUT_PREFIX, required_argument, "out-prefix" },
#if defined(XPAR_ALLOW_MAPPING)
    { FLAG_NO_MMAP, no_argument, "no-mmap" },
#endif
    { 'i', required_argument, "interlacing" },
    { 0, 0, NULL }
  };
  yarg_settings settings = { .style = YARG_STYLE_UNIX, .dash_dash = true };
  bool verbose = false, quiet = false, force = false, force_stdout = false;
  bool no_map = false, joint = false, sharded = false;
  int mode = MODE_NONE, interlacing = -1, dshards = -1, pshards = -1, jobs = -1;
  const char * out_prefix = NULL;
  yarg_result * res = yarg_parse(argc, argv, opt, settings);
  if (res->error) { fputs(res->error, stderr); exit(1); }
  for (int i = 0; i < res->argc; i++) {
    yarg_option o = res->args[i];
    switch (o.opt) {
      case 'V': version(); return 0;
      case 'j':
        if (jobs != -1) goto conflict;  jobs = atoi(o.arg); break;
      case 'J':
        if (sharded) goto conflict;  joint = true; break;
      case 'S':
        if (joint) goto conflict;  sharded = true; break;
      case 'v':
        if (quiet) goto conflict;  verbose = true; break;
      case 'q':
        if (verbose) goto conflict;  quiet = true; break;
      case 'h': help(); return 0;
      case 'e':
        if(mode == MODE_NONE) mode = MODE_ENCODING;
        else goto opmode_conflict;  break;
      case 'd':
        if(mode == MODE_NONE) mode = MODE_DECODING;
        else goto opmode_conflict;  break;
      case 'f': force = true; break;
      case 'c': force_stdout = true; break;
      case FLAG_NO_MMAP: no_map = true; break;
      case 'i':
        interlacing = atoi(o.arg);
        if (interlacing < 1 || interlacing > 3)
          FATAL("Invalid interlacing factor.");
        break;
      case FLAG_DSHARDS:
        dshards = atoi(o.arg);
        if (dshards < 1 || dshards >= MAX_DATA_SHARDS)
          FATAL("Invalid number of data shards.");
        break;
      case FLAG_PSHARDS:
        pshards = atoi(o.arg);
        if (pshards < 1 || pshards >= MAX_PARITY_SHARDS)
          FATAL("Invalid number of parity shards.");
        break;
      case FLAG_OUT_PREFIX: out_prefix = o.arg; break;
      default: exit(1); break;
      conflict: FATAL("Conflicting options.");
      opmode_conflict: FATAL("Multiple operation modes specified.");
    }
  }
#if defined(XPAR_OPENMP)
  if (jobs > 0) omp_set_num_threads(jobs);
#endif
  if (mode == MODE_NONE)
    FATAL("No operation mode specified.");
  if (!joint && !sharded) joint = true;
  if (joint) {
    if (dshards != -1 || pshards != -1 || out_prefix)
      FATAL("Sharded mode options in joint mode.");
    if (interlacing == -1) interlacing = 1;
    char * f1 = NULL, * f2 = NULL;
    switch (res->pos_argc) {
      case 0: break;
      case 1: f1 = res->pos_args[0]; break;
      case 2: f1 = res->pos_args[0], f2 = res->pos_args[1]; break;
      default: FATAL("Too many positional arguments.");
    }
    char * input_file = NULL, * output_file = NULL;
    if (f1) switch(mode) {
      case MODE_ENCODING:
        if (!f2) {
          input_file = f1;
          if (force_stdout) output_file = NULL;
          else asprintf(&output_file, "%s.xpa", f1);
        } else input_file = f1, output_file = f2;
        break;
      case MODE_DECODING:
        if (!f2) {
          input_file = f1;
          if (force_stdout) output_file = NULL;
          else {
            sz len = strlen(f1);
            if (len < 5 || strcmp(f1 + len - 4, ".xpa"))
              FATAL("Unknown file type.");
            output_file = strndup(f1, len - 4);
          }
        } else input_file = f1, output_file = f2;
        break;
    }
    joint_options_t options = {
      .input_name = input_file, .output_name = output_file,
      .interlacing = interlacing,
      .force = force, .quiet = quiet, .verbose = verbose,
      .no_map = no_map
    };
    volatile struct timeval start, end;
    gettimeofday((struct timeval *) &start, NULL);
    switch(mode) {
      case MODE_ENCODING: do_joint_encode(options); break;
      case MODE_DECODING: do_joint_decode(options); break;
    }
    gettimeofday((struct timeval *) &end, NULL);
    if (verbose) {
      double elapsed = (end.tv_sec - start.tv_sec) +
        (end.tv_usec - start.tv_usec) / 1000000.0;
      printf("Elapsed time: %.6f seconds.\n", elapsed);
    }
    if (output_file != f2) free(output_file);
  } else {
    if (interlacing != -1 || force_stdout)
      FATAL("Joint mode options in sharded mode.");
    volatile struct timeval start, end;
    gettimeofday((struct timeval *) &start, NULL);
    switch(mode) {
      case MODE_ENCODING: {
        if (dshards == -1 || pshards == -1)
          FATAL("Number of data and parity shards not specified.");
        if (res->pos_argc == 0) FATAL("No input file specified.");
        const char * input_file = res->pos_args[0];
        if (res->pos_argc > 1) FATAL("Too many positional arguments.");
        if (!out_prefix) out_prefix = input_file;
        sharded_encoding_options_t opt = {
          .input_name = input_file, .output_prefix = out_prefix,
          .dshards = dshards, .pshards = pshards,
          .force = force, .quiet = quiet, .verbose = verbose,
          .no_map = no_map
        };
        sharded_encode(opt);
        break;
      }
      case MODE_DECODING: {
        if (res->pos_argc == 0) FATAL("No output file specified.");
        const char * output_file = res->pos_args[0];
        if (res->pos_argc < 2) FATAL("No input shards specified.");
        sharded_decoding_options_t opt = {
          .output_file = output_file,
          .input_files = (const char **) res->pos_args + 1,
          .force = force, .quiet = quiet, .verbose = verbose,
          .no_map = no_map, .n_input_shards = res->pos_argc - 1
        };
        sharded_decode(opt);
        break;
      }
    }
    gettimeofday((struct timeval *) &end, NULL);
    if (verbose) {
      double elapsed = (end.tv_sec - start.tv_sec) +
        (end.tv_usec - start.tv_usec) / 1000000.0;
      printf("Elapsed time: %.6f seconds.\n", elapsed);
    }
  }
  yarg_destroy(res);
}
